### ISSUE TITLE
zarr support?

### ISSUE THREAD

**Issue by ivirshup** [theislab/zellkonverter#52]
Likely of interest to @rcannood.

If R starts to get good support for zarr (https://github.com/zarr-developers/community/issues/18), would support for reading the AnnData's written to zarr be in scope here? Particularly for a native reader/ writer as in #46 #48?

The hdf5 files are great for portability, but it's quite nice to have easier access to contemporary compression algorithms and access over http. But, it is an extra thing to maintain.

**Comment by lazappi** [comment]
I don't really know anything about zarr but I would say this is probably possible if there is some core R/Bioconductor package for handling the low-level stuff. Probably also depends on how much effort it takes to setup/maintain and if the format starts being used in practice.

For now I think we probably just leave this issue open as a reminder to keep an eye on it.

**Comment by Artur-man** [comment]
Is there still an interest in this ? Although one might implement the same utility in https://github.com/scverse/anndataR but it also lacks DelayedArray support ? 

Regardless https://github.com/grimbough/Rarr now has DelayedArray support and perhaps could be used in below like `HDF5Array` if needed:
https://github.com/theislab/zellkonverter/blob/41861c688218c23a44da20ee7d97cddc7a1985b0/R/AnnData2SCE.R#L363-L378

**Comment by Artur-man** [comment]
what would be the expected interface of `AnnData2SCE` if zarr_backed support added along with hdf5_backed, I might quickly insert an `Rarr::ZarrArray` constructor here and send in a PR.

**Comment by lazappi** [comment]
I would be happy to add Zarr support to the Python reader/write. Possibly this mostly works already, given that all the file stuff is done on the Python side. There might be some checks etc. that need adjusting but I haven't tried it to see. If someone wants to look into and make some suggestions that would be great but I'm not sure I will have time before the next release.

Any native R Zarr reading/writing would be better in **{anndataR}**. There is already a PR looking at this which people can contribute to https://github.com/scverse/anndataR/pull/190.

**Comment by Artur-man** [comment]
Yes, I think on the python side thats already available ... however what we wanna do at https://github.com/HelenaLC/SpatialData is basically to wrap a zarr dataset as DelayedArray as `AnnData` being converted to `SingleCellExperiment` instead of realizing the entire data in the memory. Currently, we capture a `table` component of a SpatialData object as SCE.  

On the other hand, and you can correct me here @lazappi, lazy reading is not available in an SCE when converted from anndata using **{anndataR}** either ? 

Here is the reader from **{SpatialData}** if you wanna check it: 
https://github.com/HelenaLC/SpatialData/blob/18a25d60beff24cc660767da78044d4055e327a0/R/read.R#L131-L137

I am open to suggestions too though. 

To me it should be an easy fix to put a `Rarr::ZarrArray` constructor under `.extract_or_skip_assay` instead. Though a combination of **{Rarr}**, **{anndataR}** and **{DelayedArray}** looks more R-native and clean (may need more work). 

I am planning to work on this a little once BioC 3.21 is on.

**Comment by lazappi** [comment]
Yes. For **{anndataR}** we don't currently have any `DelayedArray` stuff but we do a backed HDF5 wrapper which only reads things into memory as requested. It would make sense for a `DelayedArray` option to be in there somewhere. There is some support for it in **{zellkonverter}** but it's not the most tested or robust (I've never tried to use it in practice). Also, it kinda assumes the object has already been read into memory on the Python side, I'm not sure if it works if you give it a backed `AnnData` object.

I think this fits better in **{anndataR}** but I would also consider adding something here if there is a good reason to. Definitely for the next release though.

**Comment by Artur-man** [comment]
oh, I thought below checks if X is a h5py dataset but I will try now to see how it works ... 
https://github.com/theislab/zellkonverter/blob/06fe811fb8452099f0dab77dd3c020ce510013cd/R/AnnData2SCE.R#L681C1-L685C2

**Comment by Artur-man** [comment]
It seems you check the name of the h5 group in `.extract_or_skip_assay`, but `name` cannot be found however changing that to `mat$_name` fixes the issue. Otherwise, it looks like it should read an h5 dataset. You should start with `anndata::read_h5ad( ... , backed = "r")` though.

```
if (hdf5_backed && .is_anndata_matrix(mat)) {
  group_name <- as.character(py_to_r(mat$name))
  if (.h5isgroup(filepath, group_name)) {
    mat <- HDF5Array::H5SparseMatrix(filepath, group_name)
  } else {
    mat <- HDF5Array::HDF5Array(filepath, group_name)
  }
```

```
Browse[1]> mat$name
Error in py_get_attr(x, name) : 
  AttributeError: '_CSCDataset' object has no attribute 'name'. Did you mean: '_name'?
Run `reticulate::py_last_error()` for details.
Browse[1]> mat$_name
Error: unexpected input in "mat$_"
Browse[1]> mat$`_name`
'/X'
```

I have also checked `readH5AD`, that seem to work.

**Comment by Artur-man** [comment]
More information on this, as I dig deeper, currently `anndata.io.read_zarr` does not perform disk-backed reading from the zarr store either. So even before `AnnData2SCE` it will be on memory.
https://anndata.readthedocs.io/en/latest/generated/anndata.io.read_zarr.html
